sklearn/ensemble/_gb.py

```diff
@@ -23,33 +23,246 @@
+    learning_rate=0.1,
+    k=0,
+):
+    """Update the leaf values to be predicted by the tree and raw_prediction.
+
+    The current predictions of the model (of this stage) are updated.
+
+    Additionally, if loss.need_update_leaves_values is True, the terminal regions
+    (=leaves) of the given tree are updated as well. This corresponds to the line
+    search step in "Greedy Function Approximation" by Friedman.
+
+    Update equals:
+        loss.fit_intercept_only(y_true - raw_prediction)
+
+    Note: It only works, if the loss is a function of the residual, as is the
+    case for AbsoluteError and PinballLoss. Otherwise, one would need to get
+    the minimum of loss(y_true, raw_prediction + x) in x. A few examples:
+        - AbsoluteError: median(y_true - raw_prediction).
+        - PinballLoss: quantile(y_true - raw_prediction).
+    See also notes about need_update_leaves_values in BaseLoss.
+
+    Parameters
+    ----------
+    loss : BaseLoss
+    tree : tree.Tree
+        The tree object.
+    X : ndarray of shape (n_samples, n_features)
+        The data array.
+    y : ndarray of shape (n_samples,)
+        The target labels.
+    residual : ndarray of shape (n_samples,)
+        The residuals (usually the negative gradient).
+    raw_prediction : ndarray of shape (n_samples, n_trees_per_iteration)
+        The raw predictions (i.e. values from the tree leaves) of the
+        tree ensemble at iteration ``i - 1``.
+    sample_weight : ndarray of shape (n_samples,)
+        The weight of each sample.
+    sample_mask : ndarray of shape (n_samples,)
+        The sample mask to be used.
+    learning_rate : float, default=0.1
+        Learning rate shrinks the contribution of each tree by
+         ``learning_rate``.
+    k : int, default=0
+        The index of the estimator being updated.
+    """
+    # compute leaf for each sample in ``X``.
+    terminal_regions = tree.apply(X)
+
+    if not isinstance(loss, HalfSquaredError):
+        # mask all which are not in sample mask.
+        masked_terminal_regions = terminal_regions.copy()
+        masked_terminal_regions[~sample_mask] = -1
+
+        # update each leaf (= perform line search)
+        for leaf in np.nonzero(tree.children_left == TREE_LEAF)[0]:
+            indices = np.nonzero(terminal_regions == leaf)[0]  # of terminal regions
+            y_ = y.take(indices, axis=0)
+            sw = None if sample_weight is None else sample_weight[indices]
+
+            if isinstance(loss, HalfBinomialLoss):
+                # Make a single Newton-Raphson step.
+                # Our node estimate is given by:
+                #    sum(w * (y - prob)) / sum(w * prob * (1 - prob))
+                # we take advantage that: y - prob = residual
+                residual_ = residual.take(indices, axis=0)
+                prob = y_ - residual_
+                # numerator = negative gradient = y - prob
+                numerator = np.average(residual_, weights=sw)
```

Comment:
`main` uses `np.sum(sw * residual)` for the numerator. What is the reason for using average here?

(This question applies to all other places.)

Reply:
It's equivalent and better underlines the fact that when working with weights, average should be called in most circumstances, e.g. `average` instead of `mean`.